contracts/integrations/test/utils/function_assertions.ts

```diff
@@ -9,25 +12,57 @@ export interface ContractWrapperFunction extends ContractGetterFunction {
     awaitTransactionSuccessAsync?: (...args: any[]) => PromiseWithTransactionHash<TransactionReceiptWithDecodedLogs>;
 }
 
-export interface Condition {
-    before: (...args: any[]) => Promise<any>;
-    after: (beforeInfo: any, result: Result, ...args: any[]) => Promise<any>;
-}
-
 export interface Result {
     data?: any;
-    receipt?: TransactionReceiptWithDecodedLogs;
     success: boolean;
+    receipt?: TransactionReceiptWithDecodedLogs;
 }
 
-export class FunctionAssertion {
-    // A before and an after assertion that will be called around the wrapper function.
-    public condition: Condition;
+/**
+ * This interface represents a condition that can be placed on a contract function.
+ * This can be used to represent the pre- and post-conditions of a "Hoare Triple" on a
+ * given contract function. The "Hoare Triple" is a way to represent the way that a
+ * function changes state.
+ * @param before A function that will be run before a call to the contract wrapper
+ *               function. Ideally, this will be a "precondition."
+ * @param after A function that will be run after a call to the contract wrapper
+ *              function.
+ */
+export interface Condition<TBefore extends any> {
+    before: (...args: any[]) => Promise<TBefore>;
+    after: (beforeInfo: TBefore, result: Result, ...args: any[]) => Promise<any>;
+}
+
+/**
+ * The basic unit of abstraction for testing. This just consists of a command that
+ * can be run. For example, this can represent a simple command that can be run, or
+ * it can represent a command that executes a "Hoare Triple" (this is what most of
+ * our `Assertion` implementations will do in practice).
+ * @param runAsync The function to execute for the assertion.
+ */
+export interface Assertion {
+    executeAsync: (...args: any[]) => Promise<any>;
+}
+
+export interface RunResult {
+    beforeInfo: any;
+    afterInfo: any;
+}
+
+/**
+ * This class implements `Assertion` and represents a "Hoare Triple" that can be
+ * executed.
+ */
+export class FunctionAssertion<TBefore extends any> implements Assertion {
+    // A condition that will be applied to `wrapperFunction`.
+    // Note: `TBefore | undefined` is used because the `before` and `after` functions
+    //       are optional in `Condition`.
+    public condition: Condition<TBefore>;
 
     // The wrapper function that will be wrapped in assertions.
     public wrapperFunction: ContractWrapperFunction;
 
-    constructor(wrapperFunction: ContractWrapperFunction, condition: Condition) {
+    constructor(wrapperFunction: ContractWrapperFunction, condition: Condition<TBefore>) {
```

Comment:
Does this work? 
```suggestion
    constructor(wrapperFunction: ContractWrapperFunction, condition: Partial<Condition<TBefore>>) {
        this.condition = {
            before: condition.before || () => undefined,
            after: condition.after || () => undefined,
        };
```

Reply:
After discussing offline, we decided to keep this the same so that our type-safety can be as rigorous as possible